File: js/example_tts_stream.js
#!/usr/bin/env node
/**
 * Example script for Inworld TTS streaming synthesis using HTTP requests.
 *
 * This script demonstrates how to synthesize speech from text using the Inworld TTS API
 * with streaming requests, receiving audio chunks in real-time.
 */

const fs = require('fs');

/**
 * Check if INWORLD_API_KEY environment variable is set.
 * @returns {string|null} API key or null if not set
 */
function checkApiKey() {
    const apiKey = process.env.INWORLD_API_KEY;
    if (!apiKey) {
        console.log('Error: INWORLD_API_KEY environment variable is not set.');
        console.log('Please set it with: export INWORLD_API_KEY=your_api_key_here');
        return null;
    }
    return apiKey;
}

/**
 * Synthesize speech from text using Inworld TTS API with streaming.
 * 
 * @param {string} text - Text to synthesize
 * @param {string} voiceId - Voice ID to use
 * @param {string} modelId - Model ID to use
 * @param {string} apiKey - API key for authentication
 * @returns {AsyncGenerator<Buffer>} Audio chunks
 */
async function* synthesizeSpeechStream(text, voiceId, modelId, apiKey) {
    // API endpoint
    const url = 'https://api.inworld.ai/tts/v1/voice:stream';
    
    // Set up headers
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Basic ${apiKey}`
    };
    
    // Request data
    const requestData = {
        text: text,
        voice_id: voiceId,
        model_id: modelId,
        audio_config: {
            audio_encoding: 'LINEAR16',
            sample_rate_hertz: 48000
        }
    };
    
    try {
        console.log('Starting streaming synthesis...');
        console.log(`   Text: ${text}`);
        console.log(`   Voice ID: ${voiceId}`);
        console.log(`   Model ID: ${modelId}`);
        console.log();
        
        // Use native fetch API with streaming
        const response = await fetch(url, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(requestData)
        });
        
        // Check for HTTP errors
        if (!response.ok) {
            let errorDetails = '';
            try {
                errorDetails = await response.text();
            } catch {}
            throw new Error(`HTTP ${response.status}: ${errorDetails}`);
        }
        
        let chunkCount = 0;
        let totalAudioSize = 0;
        let firstChunkTime = null;
        const startTime = Date.now();
        
        console.log('Receiving audio chunks:');
        
        // Process streaming response using ReadableStream
        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        
        while (true) {
            const { done, value } = await reader.read();
            
            if (done) break;
            
            // Decode the chunk and add to buffer
            buffer += decoder.decode(value, { stream: true });
            
            // Process complete lines
            const lines = buffer.split('\n');
            buffer = lines.pop() || ''; // Keep incomplete line in buffer
            
            for (const line of lines) {
                if (line.trim()) {
                    try {
                        const chunkData = JSON.parse(line);
                        const result = chunkData.result;
                        if (result && result.audioContent) {
                            const audioChunk = Buffer.from(result.audioContent, 'base64');
                            chunkCount++;
                            totalAudioSize += audioChunk.length;
                            
                            // Record time for first chunk
                            if (chunkCount === 1) {
                                firstChunkTime = (Date.now() - startTime) / 1000;
                                console.log(`   Time to first chunk: ${firstChunkTime.toFixed(2)} seconds`);
                            }
                            
                            console.log(`   Chunk ${chunkCount}: ${audioChunk.length} bytes`);
                            yield audioChunk;
                        }
                    } catch (error) {
                        if (error instanceof SyntaxError) {
                            console.log(`   JSON decode error: ${error.message}`);
                            continue;
                        } else {
                            console.log(`   Missing key in response: ${error.message}`);
                            continue;
                        }
                    }
                }
            }
        }
        
        console.log(`\nStreaming completed!`);
        console.log(`   Total chunks: ${chunkCount}`);
        console.log(`   Total audio size: ${totalAudioSize} bytes`);
        
    } catch (error) {
        console.log(`HTTP Error: ${error.message}`);
        throw error;
    }
}

/**
 * Save streaming audio chunks to a WAV file.
 * @param {AsyncGenerator<Buffer>} audioChunks - Audio chunks generator
 * @param {string} outputFile - Output file path
 */
async function saveStreamingAudioToFile(audioChunks, outputFile) {
    try {
        console.log(`Saving audio chunks to: ${outputFile}`);
        
        // Collect all raw audio data (skip WAV headers from chunks)
        const rawAudioData = [];
        
        let i = 0;
        for await (const chunk of audioChunks) {
            // Skip WAV header if present (first 44 bytes)
            if (chunk.length > 44 && chunk.subarray(0, 4).equals(Buffer.from('RIFF'))) {
                rawAudioData.push(chunk.subarray(44));
            } else {
                rawAudioData.push(chunk);
            }
            i++;
        }
        
        const combinedAudio = Buffer.concat(rawAudioData);
        
        // Create WAV header and save file
        const wavHeader = createWavHeader(combinedAudio.length, 1, 48000, 16);
        const wavFile = Buffer.concat([wavHeader, combinedAudio]);
        
        fs.writeFileSync(outputFile, wavFile);
        
    } catch (error) {
        console.log(`Error saving audio file: ${error.message}`);
        throw error;
    }
}

/**
 * Create WAV file header.
 * @param {number} dataSize - Size of audio data
 * @param {number} channels - Number of channels
 * @param {number} sampleRate - Sample rate
 * @param {number} bitsPerSample - Bits per sample
 * @returns {Buffer} WAV header
 */
function createWavHeader(dataSize, channels, sampleRate, bitsPerSample) {
    const header = Buffer.alloc(44);
    const bytesPerSample = bitsPerSample / 8;
    const blockAlign = channels * bytesPerSample;
    const byteRate = sampleRate * blockAlign;
    
    // RIFF header
    header.write('RIFF', 0);
    header.writeUInt32LE(36 + dataSize, 4);
    header.write('WAVE', 8);
    
    // fmt chunk
    header.write('fmt ', 12);
    header.writeUInt32LE(16, 16); // chunk size
    header.writeUInt16LE(1, 20);  // audio format (PCM)
    header.writeUInt16LE(channels, 22);
    header.writeUInt32LE(sampleRate, 24);
    header.writeUInt32LE(byteRate, 28);
    header.writeUInt16LE(blockAlign, 32);
    header.writeUInt16LE(bitsPerSample, 34);
    
    // data chunk
    header.write('data', 36);
    header.writeUInt32LE(dataSize, 40);
    
    return header;
}

/**
 * Main function to demonstrate streaming TTS synthesis.
 */
async function main() {
    console.log('Inworld TTS Streaming Synthesis Example');
    console.log('=' + '='.repeat(44));
    
    // Check API key
    const apiKey = checkApiKey();
    if (!apiKey) {
        return 1;
    }
    
    // Configuration
    const text = "Hello, adventurer! What a beautiful day, isn't it?";
    const voiceId = 'Dennis';
    const modelId = 'inworld-tts-1.5-mini';
    const outputFile = 'synthesis_stream_output.wav';
    
    try {
        const audioChunks = synthesizeSpeechStream(text, voiceId, modelId, apiKey);
        await saveStreamingAudioToFile(audioChunks, outputFile);
        console.log(`Streaming synthesis completed successfully! You can play the audio file: ${outputFile}`);
        
    } catch (error) {
        console.log(`\nStreaming synthesis failed: ${error.message}`);
        return 1;
    }
    
    return 0;
}

if (require.main === module) {
    main().then(process.exit);
}

File: python/example_tts_stream.py
#!/usr/bin/env python3
"""
Example script for Inworld TTS streaming synthesis using HTTP requests.

This script demonstrates how to synthesize speech from text using the Inworld TTS API
with streaming requests, receiving audio chunks in real-time.
"""

import base64
import json
import os
import time
import wave

import requests


def check_api_key():
    """Check if INWORLD_API_KEY environment variable is set."""
    api_key = os.getenv("INWORLD_API_KEY")
    if not api_key:
        print("❌ Error: INWORLD_API_KEY environment variable is not set.")
        print("Please set it with: export INWORLD_API_KEY=your_api_key_here")
        return None
    return api_key


def synthesize_speech_stream(text: str, voice_id: str, model_id: str, api_key: str):
    """
    Synthesize speech from text using Inworld TTS API with streaming.
    
    Args:
        text: Text to synthesize
        voice_id: Voice ID to use
        model_id: Model ID to use
        api_key: API key for authentication
        
    Yields:
        bytes: Audio chunks
    """
    # API endpoint
    url = "https://api.inworld.ai/tts/v1/voice:stream"
    
    # Set up headers
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Basic {api_key}"
    }
    
    # Request data
    request_data = {
        "text": text,
        "voice_id": voice_id,
        "model_id": model_id,
        "audio_config": {
            "audio_encoding": "LINEAR16",
            "sample_rate_hertz": 48000
        }
    }
    
    try:
        print(f"🎤 Starting streaming synthesis...")
        print(f"   Text: {text}")
        print(f"   Voice ID: {voice_id}")
        print(f"   Model ID: {model_id}")
        print()
        
        with requests.post(url, headers=headers, json=request_data, stream=True) as response:
            response.raise_for_status()
            
            chunk_count = 0
            total_audio_size = 0
            first_chunk_time = None
            start_time = time.time()
            
            print("📡 Receiving audio chunks:")
            
            for line in response.iter_lines(decode_unicode=True):
                if line.strip():
                    try:
                        chunk_data = json.loads(line)
                        result = chunk_data.get("result")
                        if result and "audioContent" in result:
                            audio_chunk = base64.b64decode(result["audioContent"])
                            chunk_count += 1
                            total_audio_size += len(audio_chunk)
                            
                            # Record time for first chunk
                            if chunk_count == 1:
                                first_chunk_time = time.time() - start_time
                                print(f"   ⏱️  Time to first chunk: {first_chunk_time:.2f} seconds")
                            
                            print(f"   📦 Chunk {chunk_count}: {len(audio_chunk)} bytes")
                            yield audio_chunk
                            
                    except json.JSONDecodeError as e:
                        print(f"   ⚠️  JSON decode error: {e}")
                        continue
                    except KeyError as e:
                        print(f"   ⚠️  Missing key in response: {e}")
                        continue
            
            print(f"\n✅ Streaming completed!")
            print(f"   Total chunks: {chunk_count}")
            print(f"   Total audio size: {total_audio_size} bytes")
            
    except requests.exceptions.RequestException as e:
        print(f"❌ HTTP Error: {e}")
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_detail = e.response.json()
                print(f"   Error details: {error_detail}")
            except:
                print(f"   Response text: {e.response.text}")
        raise
    except Exception as e:
        print(f"❌ Error during streaming synthesis: {e}")
        raise


def save_streaming_audio_to_file(audio_chunks, output_file: str):
    """Save streaming audio chunks to a WAV file."""
    try:
        print(f"💾 Saving audio chunks to: {output_file}")
        
        # Collect all raw audio data (skip WAV headers from chunks)
        raw_audio_data = bytearray()
        
        for i, chunk in enumerate(audio_chunks):
            # Skip WAV header if present (first 44 bytes)
            if len(chunk) > 44 and chunk[:4] == b'RIFF':
                raw_audio_data.extend(chunk[44:])
            else:
                raw_audio_data.extend(chunk)
        
        # Save as WAV file
        with wave.open(output_file, "wb") as wf:
            wf.setnchannels(1)  # Mono
            wf.setsampwidth(2)  # 16-bit
            wf.setframerate(48000)
            wf.writeframes(raw_audio_data)
    except Exception as e:
        print(f"❌ Error saving audio file: {e}")
        raise


def main():
    """Main function to demonstrate streaming TTS synthesis."""
    print("🎵 Inworld TTS Streaming Synthesis Example")
    print("=" * 45)
    
    # Check API key
    api_key = check_api_key()
    if not api_key:
        return 1
    
    # Configuration
    text = "Hello, adventurer! What a beautiful day, isn't it?"
    voice_id = "Dennis"
    model_id = "inworld-tts-1"
    output_file = "synthesis_stream_output.wav"
    
    try:
        audio_chunks = list(synthesize_speech_stream(
            text=text,
            voice_id=voice_id,
            model_id=model_id,
            api_key=api_key
        ))
        save_streaming_audio_to_file(audio_chunks, output_file)
        print(f"🎉 Streaming synthesis completed successfully! You can play the audio file: {output_file}")
        
    except Exception as e:
        print(f"\n❌ Streaming synthesis failed: {e}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())
